internal/git/git.go

```diff
@@ -382,28 +384,146 @@ func (g *Git) CommitAndPush(openAPIDocVersion, speakeasyVersion, doc string, act
+		Message: github.String(commitMessage),
+		Tree:    &github.Tree{SHA: tree.SHA},
+		Parents: []*github.Commit{parentCommit}}, &github.CreateCommitOptions{})
 	if err != nil {
 		return "", fmt.Errorf("error committing changes: %w", err)
 	}
 
-	if err := g.repo.Push(&git.PushOptions{
-		Auth:  getGithubAuth(g.accessToken),
-		Force: true, // This is necessary because at the beginning of the workflow we reset the branch
-	}); err != nil {
-		return "", pushErr(err)
+	// Update reference
+	newRef := &github.Reference{
+		Ref:    github.String("refs/heads/" + branch),
+		Object: &github.GitObject{SHA: commitResult.SHA},
 	}
+	g.client.Git.UpdateRef(context.Background(), owner, repo, newRef, true)
 
-	return commitHash.String(), nil
+	return *commitResult.SHA, nil
 }
 
+// getOrCreateRef returns the commit branch reference object if it exists or creates it
+// from the base branch before returning it.
+func (g *Git) getOrCreateRef(commitRef string) (ref *github.Reference, err error) {
+	_, githubRepoLocation := g.getRepoMetadata()
+	owner, repo := g.getOwnerAndRepo(githubRepoLocation)
+	environmentRef := environment.GetRef()
+
+	if ref, _, err = g.client.Git.GetRef(context.Background(), owner, repo, commitRef); err == nil {
+		return ref, nil
+	}
+
+	// We consider that an error means the branch has not been found and needs to
+	// be created.
+	if commitRef == environmentRef {
+		return nil, errors.New("the commit branch does not exist but `-base-branch` is the same as `-commit-branch`")
+	}
+
+	var baseRef *github.Reference
+	if baseRef, _, err = g.client.Git.GetRef(context.Background(), owner, repo, environmentRef); err != nil {
+		return nil, err
+	}
+	newRef := &github.Reference{Ref: github.String(commitRef), Object: &github.GitObject{SHA: baseRef.Object.SHA}}
+	ref, _, err = g.client.Git.CreateRef(context.Background(), owner, repo, newRef)
+	return ref, err
+}
+
+// getTree generates the tree to commit based on the given files and the commit
+// of the ref you got in getRef.
+func (g *Git) createAndPushTree(ref *github.Reference, sourceFiles git.Status) (tree *github.Tree, err error) {
+	// Create a tree with what to commit.
+	entries := []*github.TreeEntry{}
+	_, githubRepoLocation := g.getRepoMetadata()
+	owner, repo := g.getOwnerAndRepo(githubRepoLocation)
+
+	workingDirectory, _ := g.repo.Worktree()
+
+	// Load each file into the tree.
+	for file, fileStatus := range sourceFiles {
+		if fileStatus.Staging != git.Unmodified && fileStatus.Staging != git.Untracked && fileStatus.Staging != git.Deleted {
+			filePath := workingDirectory.Filesystem.Join(workingDirectory.Filesystem.Root(), file)
+			content, err := os.ReadFile(filePath)
+			if err != nil {
+				fmt.Println("Error getting file content", err, filePath)
+				return nil, err
+			}
+			entries = append(entries, &github.TreeEntry{Path: github.String(file), Type: github.String("blob"), Content: github.String(string(content)), Mode: github.String("100644")})
```

Comment:
does Go let us shorten these line lengths or break this up?

Reply:
Good idea. I just adjusted this is so its not so long.